calico/felix/endpoint.py

```diff
@@ -472,25 +652,36 @@ def _apply_endpoint_update(self):
                         self._iptables_in_sync = False
         else:
             # Delete of the endpoint.  Need to resync everything.
-            profile_ids = set()
+            self._profile_ids_dirty = True
             self._iptables_in_sync = False
             self._device_in_sync = False
 
-        # Note: we don't actually need to wait for the activation to finish
-        # due to the dependency management in the iptables layer.
-        self.rules_ref_helper.replace_all(profile_ids)
-
         self.endpoint = pending_endpoint
         self._endpoint_update_pending = False
         self._pending_endpoint = None
 
+    def _update_profile_references(self):
+        if self.endpoint:
+            # Combine the explicit profile IDs with the set of policy IDs
+            # for our matching selectors.
+            profile_ids = set(self._explicit_profile_ids)
+            for pol_ids in self._pol_ids_by_tier.itervalues():
+                profile_ids.update(pol_ids)
+        else:
+            profile_ids = set()
+        # Note: we don't actually need to wait for the activation to finish
+        # due to the dependency management in the iptables layer.
+        self._rules_ref_helper.replace_all(profile_ids)
```

Comment:
The ref helper is just managing the set of dependencies (i.e. which profile chains we need to be present), there's no ordering needed there.  Not sure I can add much with a comment.